### ISSUE TITLE
Add collision detection

### ISSUE THREAD

**Issue by qiemem** [NetLogo/NetLogo#1407]
While basic collision detection is pretty easy in NetLogo (`any? other turtles-here`, size-distance check for circles, or bounding box checks for non-rotated rectangles), doing precise detection for other shapes is quite difficult. We should consider adding collision detection either to the core language or in an extension. Here's how it could work:

Add a TPL primitive `collides? <agent>` that checks if the current agent collides with the given agent. Here's how it would be defined for different combinations of agents:

- turtle-turtle: Use the turtle's shapes as their geometry and see if they collide. To simplify/optimize things, we could use the convex hull around the shape so we can use the separating axis theorem.
- turtle-link: Does the link pass through the turtle's shape?
- turtle-patch: The patch's geometry is just a square of size 1. Check for overlap, again, using SAT or something similar.
- link-link: Do the links intersect?
- link-patch: Does the link pass through the patch?
- patch-patch: Are they the same patch? This combination is kind of stupid, but meh.

Other possible names of the primitive:

- `touching?`
- `colliding?`
- `colliding-with?`
- ...

People could then use it like so:

```
fd speed
if any? other turtles with [ collides? myself ] [
  bk speed
]
```

Or:

```
fd speed
if any? patches with [ pcolor = grey and collides? myself ] [
  bk speed
]
```

Fancier implementations could add modified motion commands to move up until collision, but I think that's unnecessary.

**Comment by TheBizzle** [comment]
Name suggestion: `overlaps?`

**Comment by qiemem** [comment]
`overlaps?` is pretty good!

BTW, for implementer, here's a decent description of 2d collision detection of convex shapes, if we want to go that route: https://gamedevelopment.tutsplus.com/tutorials/collision-detection-using-the-separating-axis-theorem--gamedev-169

**Comment by SethTisue** [comment]
I vote for `touching?` — it's vivid and charming, `overlaps?` feels like chalkboards and textbooks

**Comment by SethTisue** [comment]
> any? patches with [ pcolor = grey and collides? myself ]

this is going to be incredibly slow, isn't it? I wonder if this is a case like `in-radius` and `distance` where technically `in-radius` could be defined in terms of distance, but it's faster since it doesn't need to check every agent.

not sure what the noun-phrase primitive name would be — `touching-turtles`, `touching-patches`, `touching-<breed>` ...?

**Comment by qiemem** [comment]
> this will be super slow, won't it? perhaps this is a case like in-radius and distance where technically in-radius could be defined in terms of distance, but it's faster since it doesn't need to check every agent.

Yes, I was just giving a simple example. If it does turn out to be slow in a users particular model, they could do something like:

```
any? patches in-radius (0.75 * size + 1) with [ pcolor = grey and touching? myself ]
```
The 0.75 gives you approximately the bounding circle of the turtle. I think I would favor documenting such optimizations rather than implementing additional primitives for them, but I could be convinced otherwise.

**Comment by UriCCL** [comment]
In various dialects of Logo, touching? is a primitive. I’ve always liked it.

—Uri

Uri Wilensky

**Comment by fstonedahl** [comment]
While it sounds cool, a counterargument would be that this is a further break-down of the view/model boundary.  (We already do this a fair amount with colors, where we use the color of a patch/turtle to signify model state, so maybe that ship has sailed, but...) 

At present, modelers should understand that turtles are modeling specific *point* locations.  Size and shape properties are merely for visual effect, and do not (usually) impact how the model runs.  While this paradigm is less powerful than giving turtles full-blown geometry, it is conceptually simpler, and it is consistent.  Certainly, if someone updated the graphics on the sheep shape later on, I wouldn't want the Wolf-Sheep model to behave any differently.

The efficiency concerns are also worth considering, especially as novice modelers might use touching? without realizing the performance penalty that will come with it.

There are also some design choices about corner cases -- like would we agree that a patch is  "**touching?**" its **neighbors**/**neighbors4**?   If a link has **thickness**, does it intersect as a rotated rectangle instead of as a line?  What about curved links -- will the mathematics be reasonable for computing those intersections?   Also, you need the math to on all world topologies... so it maybe a somewhat large undertaking.

All that said -- I'm not necessarily against this feature (I think it would be fun to play with), but I'd be curious to hear some more solid use cases of models (either educational or scientific) that would really take advantage of this (and be the better for it).  

If it can be implemented in an extension, then I don't see any big downsides, since people can choose whether to use it or not.

**Comment by fstonedahl** [comment]
P.S.  One more design consideration -- what if a turtle or link is hidden?  Does it still touch things?

**Comment by christiancrisostomo** [comment]
Hi! Finally, the collision detector was added?

Thanks

**Comment by brandesNW** [comment]
Hi,
I’m part of the development team and I haven’t heard anything about work on collision detection.

  *   Aaron

--
Aaron Brandes, Software Developer
Center for Connected Learning and Computer-Based Modeling


From: Christian Torres-Sosa ***@***.***>
Date: Monday, March 6, 2023 at 9:45 AM
To: NetLogo/NetLogo ***@***.***>
Cc: Subscribed ***@***.***>
Subject: Re: [NetLogo/NetLogo] Add collision detection (#1407)

Hi! Finally, the collision detector was added?

Thanks

**Comment by LaCuneta** [comment]
@christiancrisostomo  While full shape-based collision detection has not been completed, there is now [a Physics Extension for NetLogo](https://github.com/Loafie/netlogo_phys_extension) that can detect collisions between circular shapes and also flat patch "walls".  Depending on what you're trying to do that might be enough functionality.  It's available in the extensions manager for easy installation.

**Comment by christiancrisostomo** [comment]
That was my doubt. Thank you very much for your answer!

**Comment by SinnnwL** [comment]
> @christiancrisostomo While full shape-based collision detection has not been completed, there is now [a Physics Extension for NetLogo](https://github.com/Loafie/netlogo_phys_extension) that can detect collisions between circular shapes and also flat patch "walls". Depending on what you're trying to do that might be enough functionality. It's available in the extensions manager for easy installation.

@LaCuneta,

Yes, the Physics Extension appears to be quite useful. However, I'm doubtful that it's feasible to maintain collisions between our turtles when employing the basic move functions in NetLogo (fd, bk).

Unless there is another method available to achieve this objective.
